Playwright: migrate theme-activate spec.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR migrates the `wp-theme__activate-spec` from Selenium.

Key changes:

- new method to generate a random integer between bounds in `DataHelper`. This is used to select a random theme if more than 1 matching themes are found after a keyword search.
- new class `ThemeCustomizerPage` to represent the `Appearance > Customize` page.

#### Testing instructions

TeamCity
- ensure that Playwright E2E tests pass.
- ensure that unit tests and other tasks pass.
- ensure that for Playwright E2E tests, that Theme related tests are not muted.

Local
- pull changes
- transpile with `yarn workspace @automattic/calypso-e2e build --watch` 
- run the following:
```
DEBUG=pw:api mocha --config .mocharc_playwright.yml specs/specs-playwright/wp-theme__switch-activate-spec.js
```

Blocked by #54006

worldomonation committed:
Implement skeleton of theme switch: activate test.

worldomonation committed:
Implement new test `theme - switch and activate`

worldomonation committed:
Correct parameter type

worldomonation committed:
Refactor how the theme is selected in case there is a fuzzy selection.

worldomonation committed:
- Add JSDoc to new classes and methods.
- add optional parameter to ThemesDetailPage.activate that allows caller to keep the thanks message modal open.
- modify how the ThemesCustomizerPage popup is handled from the ThemesDetailPage.

worldomonation committed:
Specify subitem in sidebar.gotoMenu.

worldomonation committed:
More comments in a tricky part of the code.

worldomonation commented on packages/calypso-e2e/src/lib/pages/themes-customizer-page.ts:
```diff
@@ -0,0 +1,44 @@
+	 * @returns {Promise<void>} No return value.
+	 */
+	async closeTab(): Promise< void > {
+		await this.page.close();
```
This is the Customizer page, and since we aren't doing anything with this page at this point in time, this class is quite basic and only checks the `iframe` loads.

dpasque replied:
What do you think of potentially renaming this class/file to be the `ThemesCustomizerPageWrapper` or something like that?

Because I believe if you go Appearance > Customize (like you say in the comments), there actually is _not_ an iframe - we only use that iframe in the popout tab that launches when you go from the theme selector.

That way, we could differentiate this wrapper holding that iframe from actual Customizer page content itself (which will probably have a lot of different actions on it actually related to customizing).

worldomonation replied:
This is some hard hitting investigative journalism here!

As I mentioned in the comment thread above - I had wrongly assumed that every instance of this customizer would be wrapped in an iFrame. This is totally my oversight and a bad assumption to boot, so thank you for investigating this.

I did think a bit over lunch on how best to go forward on this.

Since the presence of this object isn't really a critical piece of the actual spec, I am inclined to remove this class over renaming this to a wrapper. This would require modifying the test to wait on a popup, wait for it to load then perhaps check the title, all in the spec itself.

What do you think? I am not opposed to keeping the wrapper class with a new name either :)

dpasque replied:
Yes, I really like the suggestion of removing the class, great idea! It's a pretty thin wrapper we have to interact with, and I think interacting with that pop-up tab and iframe directly, if needed, is very reasonably within the purview of a test script as it is needed.

worldomonation commented on packages/calypso-e2e/src/lib/pages/themes-detail-page.ts:
```diff
@@ -23,4 +33,40 @@ export class ThemesDetailPage extends BaseContainer {
+		await this.page.click( selectors.activateModalButton );
+		await this.page.waitForSelector( selectors.thanksMessage );
+		if ( ! keepModal ) {
+			await this.page.keyboard.press( 'Escape' );
```
I chose to use the ESCAPE key to dismiss the modal instead of alternative (eg. clicking on the overlay outside of the modal). 

Hard to say what users do more often, but I suspect this isn't too important in the grand scheme of things.

dpasque replied:
I like that! It seems a cleaner, more stable option than trying to find the right spot in the overlay to click for sure.

worldomonation commented on packages/calypso-e2e/src/lib/pages/themes-detail-page.ts:
```diff
@@ -23,4 +33,40 @@ export class ThemesDetailPage extends BaseContainer {
+	async customizeSite(): Promise< Page > {
+		const [ popup ] = await Promise.all( [
+			this.page.waitForEvent( 'popup' ),
+			this.page.click( selectors.customizeDesignButton ),
```
Clicking on the `Customize site` span at top of the page **_after_** the theme has been applied will bring up the customizer in a popup.

worldomonation commented on packages/calypso-e2e/src/lib/pages/themes-page.ts:
```diff
@@ -61,28 +61,42 @@ export class ThemesPage extends BaseContainer {
+		await selectedTheme.hover();
+		await selectedTheme.waitForElementState( 'stable' );
+
+		await Promise.all( [ this.page.waitForNavigation(), selectedTheme.click() ] );
```
This method sees a large refactor in this PR. 

Code in the previous form was leading to tests that failed to successfully click the target theme (Line 86) which led to rest of the script stalling due to this unexpected behavior.

This revised code should achieve better reliability when clicking on the target theme, while adding ability to accept  fuzzy names.

worldomonation replied:
#54613 is the task to investigate flakiness in the original implementation of this code. It _should_ be resolved with this patch.

dpasque replied:
I really like the new selector you're using - I think that should definitely help with the false positives. 

Do you think there's any danger in adding the random element to this? As in, does that make the test less declarative and harder to diagnose if something goes wrong? 

I think the fuzzy matching is a great call, because it gives protection against future theme addition (and makes it easier to use). But should we just grab the first available theme by index?

The scenario I'm imagining is that someone writes a test and is expecting to have been switched to a specific theme (maybe because it's the only one that matched before). Now, the search returns more results. I think they would see the test randomly pass and fail some times, and that might complicate investigation.

I don't feel strongly about this, just wanted to throw out the devils advocate piece!

worldomonation replied:
Definitely a super good point you raise!

The [original Selenium spec](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/specs/specs-calypso/wp-theme__switch-activate-spec.js#L34-L41) and the [themes-page](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/lib/pages/themes-page.js#L57-L60) that implement the page object behave like the way you suggest: it essentially cycles between two themes constantly. 

It is probably worth refactoring this part to add determinism by replicating the Selenium spec's behavior.

worldomonation replied:
I've modified this section as follows:

- no longer use the generated pseudo-random value `index` in determining what theme is targeted for selection. 

The change was quite simple; Playwright's `nth-match` uses a 1-indexed match, and the `selector` was already only matching non-active themes, so it was only a matter of swapping out the integer value with a fixed value `1`.

worldomonation committed:
Switch theme activate spec to Jest.

worldomonation committed:
Merge branch 'trunk' into add/playwright-theme-acticate

dpasque commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -2,6 +2,17 @@ import phrase from 'asana-phrase';
+ * @param {number} max Maximum value.
+ * @returns {number} Generated pseudo-random integer.
+ */
+export function getRandomInteger( min = 0, max: number ): number {
```
This is a bit of a style nitpick, but in my experience having optional parameters before required in the signatures can sometimes feel awkward to the callers because they have to pass in null or undefined to use that optional parameter:
```typescript
getRandomInteger(undefined, 8)
```
One option would be to just move the `min` parameter towards the end!

Alternatively, I actually think it wouldn't be so bad to force users to specify the range to make the code even more explicit. E.g.,  a signature kind of like:

```typescript
getRandomIntegerInRange(min: number, max:number): number
```
which makes the nice read of:
```typescript
getRandomIntegerInRange(1, 8);
```

worldomonation replied:
Agreed on both counts of parameter and function naming. I will push up a revision with changes later today.

dpasque commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -2,6 +2,17 @@ import phrase from 'asana-phrase';
+ * @returns {number} Generated pseudo-random integer.
+ */
+export function getRandomInteger( min = 0, max: number ): number {
+	return Math.floor( Math.random() * ( max - min + 1 ) ) + min;
```
The [MDN page example for this kind of function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random#getting_a_random_integer_between_two_values_inclusive) includes using `ceil` and `floor` to protect against people passing in bad data. This protects against the edge case that can actually create a number outside the range. (E.g. if the min was `2.5`, and Math.random() rolled a `0`, the result would be `2`, outside of the range).

What do you think about adding that here?

worldomonation replied:
Good point, will be revised in next push.

dpasque commented on packages/calypso-e2e/src/data-helper.ts:
```diff
@@ -2,6 +2,17 @@ import phrase from 'asana-phrase';
 import { getViewportName } from './browser-helper';
 
+/**
+ * Generate a pseudo-random integer.
```
What do you think of adding a small comment about this being inclusive on both ends (minimum and maximum)? That might be useful for consuming callers to know!

dpasque commented on packages/calypso-e2e/src/lib/pages/themes-customizer-page.ts:
```diff
@@ -0,0 +1,44 @@
+	 *
+	 * @returns {Promise<void>} No return value.
+	 */
+	async _postInit(): Promise< void > {
+		const elementHandle = await this.page.waitForSelector( selectors.frame );
+		this.frame = ( await elementHandle.contentFrame() ) as Frame;
+	}
```
Quick clarification: are we using `_postInit` here instead of passing that iframe selector into the constructor in order to stash off a pointer to the iframe?

I just want to make sure I am understanding the difference between the constructor and the selector passed there to be used during `Expect > _init()`, and `_postInit`

worldomonation replied:
Good observation! 

I am using `_postInit` to indeed save the page element referenced by the selector. This doesn't have to be done in `_postInit`, but I had wrongly assumed that all instances of the Customizer page would be wrapped in an iFrame. 

So this `_postInit` may not stay in the next revision, depending on how I decide to implement the page object for Customizer that can work with both iFrame wrapped and regular page instances.

worldomonation committed:
ESLint fixes.

dpasque commented on packages/calypso-e2e/src/lib/pages/themes-customizer-page.ts:
```diff
@@ -0,0 +1,44 @@
+ * @augments {BaseContainer}
+ */
+export class ThemesCustomizerPage extends BaseContainer {
+	frame!: Frame;
```
Since we can now with Typescript, should this be a private class member? I imagine this variable being used by various methods in this class, but probably not as much by calling classes. Do you think it will be needed by the consuming tests?

worldomonation replied:
You are right - these are internal class member and do not need to be accessed externally. It would make sense to have these members (`frame`, `page`, `block`, etc.) be private in addition to the `_postInit`, `_preInit` and similar methods.

I will start with making the members in this class private/protected first.

dpasque replied:
Class no longer exists, so just resolving this comment!

dpasque commented on packages/calypso-e2e/src/lib/pages/themes-detail-page.ts:
```diff
@@ -20,4 +31,40 @@ export class ThemesDetailPage extends BaseContainer {
+	 */
+	async activate( { keepModal = false }: { keepModal?: boolean } = {} ): Promise< void > {
+		await this.page.click( selectors.activateDesignButton );
+		await this.page.waitForSelector( selectors.activateModal );
```
I read some suggestions from Bart recently that proposed taking out any waiting for elements that won't actually be interacted with directly to add more stability to the test.

In that vein, what do you think of taking out this extra wait for the activateModal? The only thing we really need is the button in the next line (at least it seems to me).

worldomonation replied:
Good spotting. Will remove in next push.

dpasque reviewed (commented):
GitHub is freaking out on me in a weird way, so I'm going to submit these comments real quick as to not lose them, and keep looking at a few other changes!

dpasque commented on packages/calypso-e2e/src/lib/pages/themes-page.ts:
```diff
@@ -58,28 +58,42 @@ export class ThemesPage extends BaseContainer {
+		// currently activated themes.
+		const selector = `${ selectors.items }:has-text("${ name }")${ selectors.excludeActiveTheme }`;
+		// Get number of themes being shown on page.
+		const numThemes = await this.page.$$( selector ).then( ( themes ) => themes.length );
```
I might be misunderstanding how the empty case is handled here, but do we want to through if numThemes is 0?

I worry that the behavior of trying to use the index of 0 later in the waitForSelector might result in a more confusing error message than one that explicitly says something like "No themes were found matching the search name".

worldomonation replied:
Good point. I tried searching for a theme with name `wejrlwerjlkwejr` which would of course return no results.

I will add a case to handle if `numThemes === 0` to throw.

worldomonation replied:
Added case to handle if `numTheme === 0`.

dpasque commented on packages/calypso-e2e/src/lib/pages/themes-page.ts:
```diff
@@ -58,28 +58,42 @@ export class ThemesPage extends BaseContainer {
 	async search( keyword: string ): Promise< void > {
 		const searchInput = await this.page.waitForSelector( selectors.searchInput );
-		await searchInput.fill( keyword );
+		await Promise.all( [ this.page.waitForNavigation(), searchInput.fill( keyword ) ] );
```
Similar to my other question, do we want to wait for navigation specifically _after_ we are done filling in the search keyword? I.e.m moving the `await this.page.waitForNaviation()` outside the `Promise.all()`?

worldomonation replied:
This is a [fairly common](https://playwright.dev/docs/navigations) pattern in Playwright:
```
await Promise.all([
    wait for navigation
    something that triggers navigattion
])
```

This is because entering text in the Themes search box triggers a navigation. 

By placing `waitForNavigation` prior to the action, Playwright begins observing for a navigation and this is fulfilled when the action fires next. This can be observed in the logs that show what the Playwright API is doing in the background:
```
  pw:api => page.waitForNavigation started +1ms
  pw:api => elementHandle.fill started +2ms
  pw:api waiting for navigation until "load" +3ms
  pw:api elementHandle.fill("Twenty Twen") +1ms
  pw:api   waiting for element to be visible, enabled and editable +3s
  pw:api   element is visible, enabled and editable +17ms
  pw:api <= elementHandle.fill succeeded +12ms
  pw:api   navigated to "https://wordpress.com/themes/free/e2eflowtesting3.wordpress.com?s=Twenty+Twen" +331ms
  pw:api <= page.waitForNavigation succeeded +3ms
```

In this specific case here, I grouped the `fill` and `waitForNavigation` as a way to reduce reliance on selectors when waiting for actions on the page to be completed.

dpasque reviewed (commented):
Script migration looks great and really easy to read! 😃 

Just have a few more suggestions and questions to take a look at before 🚢 .

worldomonation committed:
Address Dan's PR review.
- discontinue use of pseudo-randomly generated number in selecting the target theme to switch to.
- remove unnecessary `waitForSelector`.
- remove ThemeCustomizerPage due to invalid assumptions held on the structure of the page. Instead, modify the test steps to confirm loading of the page then close.

worldomonation committed:
Remove unused import.

dpasque commented:
@worldomonation All the changes look fantastic! This script transition landed in a really nice spot. 🚢  away!!
